common/djangoapps/third_party_auth/api/views.py

```diff
@@ -22,13 +23,42 @@
 from third_party_auth.provider import Registry
 
 
+class ProviderBaseThrottle(throttling.UserRateThrottle):
+    """
+    Base throttle for provider queries
+    """
+
+    def allow_request(self, request, view):
+        """
+        Only throttle unauthenticated requests.
+        """
+        if view.is_unprivileged_query(request, view.kwargs[u'username']):
+            return True
+        return super(ProviderBaseThrottle, self).allow_request(request, view)
+
+
+class ProviderBurstThrottle(ProviderBaseThrottle):
+    """
+    Maximum number of provider requests in a quick burst.
+    """
+    rate = '10/min'
+
+
+class ProviderSustainedThrottle(ProviderBaseThrottle):
+    """
+    Maximum number of provider requests over time.
+    """
+    rate = '50/day'
+
+
 class UserView(APIView):
     """
     List the third party auth accounts linked to the specified user account.
 
     **Example Request**
 
         GET /api/third_party_auth/v0/users/{username}
+        GET /api/third_party_auth/v0/users/{email@example.com}
```

Comment:
Can you explain why the username identifier is made into an ambiguous identifier here?  In general, I think we were trying to phase out ambiguous identifiers whenever possible, so I have to ask when we're adding new ones.

Reply:
Phase it out in favor of what?  

Our client that we did the downstream work for had a requirement of allowing users to log in by either identifier. 

Would you prefer to have different endpoints for each, or to use a query parameter to identify the type of identifier?

Reply:
Thanks @pwnage101.  That's helpful information.

FYI @xitij2000 and @bradenmacdonald: This is related to the SSO provider lookup we did recently.  This would require a minor change and the client side (having the api client choose an identifier type before making the request) and the server side (providing the argument as a query parameter) in the downstream code.  Can we do that work to help edx move forward with unambiguous user identifiers?

Reply:
We can't immediately change the API contract, because it would break the existing apps, but if we did want to push the "username vs. email" logic into the clients and make this API optionally unambiguous (by requiring a parameter/path to explicitly specify email vs. username), and later require it to be unambiguous, that would be fine.

Reply:
I'll chime in here in summary: For this PR to proceed, we'd very much like it to include functionality to allow clients to unambiguously state whether they're sending an email address or a username, whether it's by two different new URLs -or- optional parameters on this endpoint. The old ambiguous endpoint would remain functional for now until clients can be shifted to use the unambiguous endpoint. But there'd be no remaining blocker to shifting clients over to use the unambiguous endpoint.

After clients are shifted, we'd then remove the old ambiguous endpoint functionality - perhaps in a new API version, if needed. Thoughts?

Reply:
Sounds good to me.  Here's how I'll implement that:

The proposed endpoint will remain.  A new route will be added at `/api/third_party_auth/v0/users/` (with no path segment for the ambiguous identifier), that takes one of `?username=` or `?email=`.  If neither is provided, the endpoint will return a 400.